### PR TITLE
gbm-kms: Add “--driver-quirks” option

### PR DESCRIPTION
**Pull request by RAOF**
There exist devices where merely running our `probe_graphics_driver()` routine will crash.

Add a `gbm-kms`-specific option `--driver-quirks` to allow the user to specify device-specific quirks.
Currently, only quirks of the form `skip:(devnode|driver):$VALUE` are supported. For example,
`skip:devnode:/dev/dri/card0` will add a quirk refusing to probe or use the `/dev/dri/card0` device,
and `skip:driver:amdgpu` will add a quirk refusing to probe or use any device using the `amdgpu` driver.

### CODE REVIEW

**Comment by RAOF** [comment]
I've done this in a way that the `Quirks` class *should* be reasonably easily extracted to a platform-level interface if we need to.

There are some obvious extensions that could be made, particularly loading quirks from a file. The current format of the option should allow that to be done without disruption; `--driver-quirks=file:/path/to/file` wouldn't conflict with any existing quirk.

File: src/platforms/gbm-kms/server/kms/quirks.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    "Ignoring unexpected value for %s option: %s (expects value of the form “skip:<type>:<value>”)",
+                    quirks_option_name,
+                    quirk.c_str());
+                continue;
+            }
+            if (quirk[0] != 's' ||
+                quirk[1] != 'k' ||
+                quirk[2] != 'i' ||
+                quirk[3] != 'p' ||
+                quirk[4] != ':')
```
Comment (AlanGriffiths) [inline src/platforms/gbm-kms/server/kms/quirks.cpp:78]: `strncmp()`?

File: tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright © 2012 Canonical Ltd.
```
Comment (AlanGriffiths) [inline tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp:2]: s/12/21/

File: tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Alexandros Frantzis <alexandros.frantzis@canonical.com>
```
Comment (AlanGriffiths) [inline tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp:16]: Really?

File: tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+TEST(Quirks, skip_devnode_quirk_fires)
+{
+    mtf::UdevEnvironment udev;
+    udev.add_standard_device("standard-drm-devices");
+
+    boost::program_options::options_description description;
+    mir::options::ProgramOption options;
+
+    mgg::Quirks::add_quirks_option(description);
```
Comment (AlanGriffiths) [inline tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp:41]: Why not extract the common setup into a test fixture?

**Review by AlanGriffiths — APPROVED** [review]
Seems reasonable, but a few nits.

bors r=@RAOF

**Comment by AlanGriffiths** [comment]
bors delegate=@RAOF

File: tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp (hunk 3 above)
Comment (RAOF) [reply tests/unit-tests/platforms/gbm-kms/kms/test_quirks.cpp:16]: Oops, did I not update that copyright header :blush:

File: src/platforms/gbm-kms/server/kms/quirks.cpp (hunk 1 above)
Comment (RAOF) [reply src/platforms/gbm-kms/server/kms/quirks.cpp:78]: That would indeed be the function I was hunting for, thanks.

**Review by AlanGriffiths — APPROVED** [review]
bors r+

**Pull request merged**